Kernel/Syscalls/setpgid.cpp

```diff
@@ -124,7 +124,7 @@ int Process::sys$setpgid(pid_t specified_pid, pid_t specified_pgid)
     ProcessGroupID new_pgid = specified_pgid ? ProcessGroupID(specified_pgid) : process->m_pid.value();
     SessionID current_sid = sid();
     SessionID new_sid = get_sid_from_pgid(new_pgid);
-    if (current_sid != new_sid) {
+    if (new_sid != -1 && current_sid != new_sid) {
```

Comment:
I remain unsure whether this is the right thing to do, if the `pgid` doesn't exist, there will be no "new session" -- is `setpgid()` only allowed for valid, existing `pgid`'s, or can it also create a new group?

Reply:
Dr. POSIX says this is the correct behaviour (https://pubs.opengroup.org/onlinepubs/9699919799/functions/setpgid.html), and I am apparently very blind.

Reply:
What I wanted to achieve while changing `get_sid_from_pid` is that it never "accidentally" returns -1. Also, 

`current_sid` should *never* be -1, no matter what happens. So if `new_sid==-1` (which should only happen with "invalid-ish" pgids), then `current_sid != new_sid` already catches this.

Did you ever observe a `sid` of -1? If so, we should probably add an assert there, and fix that issue first.

Reply:
Btw, I like using this website, as that's what kling also uses: https://linux.die.net/man/3/setpgid

Reply:
*coin drops*

I've been misreading this diff the whole time. This diff is 100% correct. I introduced a bug by disallowing to change to a new session ID! :o

Reply:
Actually, now that I think about it, this is weird.
- Creating a new PGID should only be allowed if the new PGID=PID
- However, POSIX seems to be silent on this issue.

I'm strongly in favor of adding the requirement, because otherwise a malicious process (even a different user `nona`!) can guess PGIDs and "block" them, thus interfering with `setpgid` of `anon`'s processes.

I suggest implementing it like so:

```
    if (your previous expression)
        return -EPERM;
    if (new_sid == -1 && new_pgid != pid)
        return -EPERM;
```

Reply:
That sounds reasonable to me.